numpy/core/src/multiarray/compiled_base.c

```diff
@@ -500,6 +506,37 @@ binary_search(double key, double arr [], npy_intp len)
     if (key > arr[len - 1]) {
         return len;
     }
+    else if (key < arr[0]) {
+        return -1;
+    }
+
+    if (guess < 0)  {
+        guess = 0;
+    }
+    else if (guess >= len - 1) {
+        guess = len - 2;
+    }
+
+    /* check most likely values: guess, guess + 1, guess - 1 */
```

Comment:
@charris  I don't understand the comment about "before line:513" -- 
The code between line 513:518 ensures `guess` is in range.

Re: redundancies (I think only access of `arr[guess]`, `arr[guess + 1]` and `arr[guess - 1]`).  I thought this did not warrant making local variables, and to let the compiler optimize this.

The code between line 532:538 could be removed -- it's probably too machine-dependent, and seems to help only for a limited kind of "not well-ordered" data.   Otherwise, it's not likely to do any harm, but that might be faint praise.